### PR TITLE
Modify code base to support the addition of the IDP_ID column to token table

### PR DESCRIPTION
**Pull request by janakamarasena**
### Proposed changes in this pull request

This PR modifies the code base to support the addition of the IDP_ID column to the token table.
Includes query changes, cache key changes and IDP_ID column availability check config.

Queries for schema changes
```
ALTER TABLE IDN_OAUTH2_AUTHORIZATION_CODE
ADD COLUMN IDP_ID INTEGER AFTER AUTHORIZATION_CODE_HASH;

ALTER TABLE IDN_OAUTH2_ACCESS_TOKEN
ADD COLUMN IDP_ID INTEGER AFTER REFRESH_TOKEN_HASH;

ALTER TABLE IDN_OAUTH2_ACCESS_TOKEN 
DROP INDEX CON_APP_KEY,
ADD CONSTRAINT CON_APP_KEY UNIQUE (CONSUMER_KEY_ID,AUTHZ_USER,TENANT_ID,
USER_DOMAIN,USER_TYPE,TOKEN_SCOPE_HASH,TOKEN_STATE,TOKEN_STATE_ID,IDP_ID);

ALTER TABLE IDN_OAUTH2_ACCESS_TOKEN_AUDIT
ADD COLUMN IDP_ID INTEGER AFTER INVALIDATED_TIME;

```
<br><br><br><br><br><br><br><br><br>

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

**Comment by janakamarasena** [comment]
Travis is failing with SQL exceptions due to the [isIDPIdColumnEnabled() config](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1097/commits/a9394ab4e35636142ef8cc1b26749b46c9fe06c0#diff-63bd03c7642c088e8b6644c7a8187a4cR91) is hardcoded to true ATM but the IDP_ID column is unavailable in the relevant tables ATM.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java [hunk 1]
```diff
@@ -86,6 +86,11 @@ public static void setAudienceEnabled(boolean audienceEnabled) {
         OAuth2ServiceComponentHolder.audienceEnabled = audienceEnabled;
     }
 
+    public static boolean isIDPIdColumnEnabled() {
+        // TODO:#FIDP-FIX: Add proper config
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java]: Please do not have TODO.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAOImpl.java [hunk 2]
```diff
@@ -161,18 +180,13 @@ public RefreshTokenValidationDataDO validateRefreshToken(String consumerKey, Str
...
-                                    "Hence setting user to a federated user for client id" + consumerKey);
-                        }
-                        user.setFederatedUser(true);
+                    String authenticatedIDP = null;
+                    if (OAuth2ServiceComponentHolder.isIDPIdColumnEnabled()) {
+                        authenticatedIDP = resultSet.getString(12);
                     }
+                    AuthenticatedUser user = OAuth2Util.createAuthenticatedUser(userName, userDomain, tenantDomain,
+                            authenticatedIDP);
+                    user.setAuthenticatedSubjectIdentifier(subjectIdentifier);
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAOImpl.java]: This is the only place I noted that subjectIdentifier is being set explicitly.
It should be set in each case right as being retrieved from the DB or the same create method should handle it

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenManagementDAOImpl.java]: This line of code was actually previously [there](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1097/files/ba51e19b7c89ebd9e8a105186974e25b284498eb#diff-88d0d684bd0bd2fb1bc1b5bd8f26f48cL165). There are also several other places where this is being used.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OldTokensCleanDAO.java [hunk 3]
```diff
@@ -84,7 +102,15 @@ public void cleanupTokenByTokenValue(String token, Connection connection) throws
...
-        PreparedStatement insertintoaudittable = connection.prepareStatement(SQLQueries.STORE_OLD_TOKEN_IN_AUDIT);
+
+        String sql;
+        if (OAuth2ServiceComponentHolder.isIDPIdColumnEnabled()) {
+            sql = SQLQueries.STORE_OLD_TOKEN_IN_AUDIT_WITH_IDP_NAME;
+        } else {
+            sql = SQLQueries.STORE_OLD_TOKEN_IN_AUDIT;
+        }
+
+        PreparedStatement insertintoaudittable = connection.prepareStatement(sql);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OldTokensCleanDAO.java]: please change the variable 'insertintoaudittable' to camel case.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 4]
```diff
@@ -2974,4 +2981,77 @@ public static void addTokenDOtoCache(AccessTokenDO newTokenBean) throws Identity
...
+     * @param user Authenticated User.
+     * @return Returns the sanitized user store domain.
+     */
+    public static String getUserStoreDomain(AuthenticatedUser user) {
+
+        String userDomain;
+        if (OAuth2ServiceComponentHolder.isIDPIdColumnEnabled() &&
+                !OAuthServerConfiguration.getInstance().isMapFederatedUsersToLocal() && user.isFederatedUser()) {
+            if (log.isDebugEnabled()) {
+                log.debug("IDP_ID column is available. User is federated and not mapped to local users.");
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: shall we log some context information as well? e.g. username and tenant domain

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Improved the logs. Added user.toString() which will contain the fully-qualified username.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java (hunk 1 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java]: removed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OldTokensCleanDAO.java (hunk 3 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OldTokensCleanDAO.java]: There are few more code to be refactored related to old token cleanup. I have created this issue to track it https://github.com/wso2/product-is/issues/5195

**Comment by janakamarasena** [comment]
Commit 1c8cb56 modifies the unit tests that were affected by the addition of the IDP_ID column to run the tests with and without the IDP_ID column changes.
Also modified the OAuth cache keys used in unit tests to reflect the new cache keys

**Pull request merged**